File: app/src/main/java/com/byoutline/kickmaterial/model/DiscoverQuery.java
package com.byoutline.kickmaterial.model;

import android.databinding.BaseObservable;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Sebastian Kacprzak <sebastian.kacprzak at byoutline.com>
 */
public class DiscoverQuery extends BaseObservable {
    @Nonnull
    public final Map<String, String> queryMap;
    @Nonnull
    public final DiscoverType discoverType;
    private static final int PER_PAGE = 12;

    public DiscoverQuery(@Nonnull Map<String, String> queryMap, @Nonnull DiscoverType discoverType) {
        this.queryMap = queryMap;
        this.discoverType = discoverType;
    }

    public Integer getPageFromQuery() {
        if(queryMap != null && queryMap.containsKey("page")) {
            return Integer.valueOf(queryMap.get("page"));
        }
        return null;
    }

    public static DiscoverQuery getDiscover(@Nullable Integer page) {
        boolean firstPage = page == null || page == 1;
        Map<String, String> params = firstPage ? Collections.emptyMap() : getDiscoverCategoryMap(null, page, PER_PAGE, null);
        return new DiscoverQuery(params, DiscoverType.DISCOVER);
    }

    public static DiscoverQuery getDiscoverMore(Map<String, String> params) {
        return new DiscoverQuery(params, DiscoverType.DISCOVER_MORE);
    }

    public static DiscoverQuery getDiscoverCategory(int categoryId, int page, @Nonnull SortTypes sort) {
        Map<String, String> params = getDiscoverCategoryMap(categoryId, page, PER_PAGE, sort);
        return new DiscoverQuery(params, DiscoverType.DISCOVER_CATEGORY);
    }

    public static DiscoverQuery getDiscoverSearch(@Nonnull String searchTerm, @Nullable Integer categoryId, @Nullable Integer page, @Nonnull SortTypes sort) {
        Map<String, String> params = getDiscoverCategoryMap(categoryId, page, PER_PAGE, sort);
        params.put("term", searchTerm);
        return new DiscoverQuery(params, DiscoverType.SEARCH);
    }

    private static Map<String, String> getDiscoverCategoryMap(@Nullable Integer categoryId, @Nullable Integer page, @Nullable Integer perPage, @Nullable SortTypes sort) {
        Map<String, String> params = new HashMap<>();
        if (categoryId != null) {
            params.put("category_id", Integer.toString(categoryId));
        }
        if (page != null) {
            params.put("page", Integer.toString(page));
        }
        if (perPage != null) {
            params.put("per_page", Integer.toString(perPage));
        }
        if (sort != null) {
            params.put("sort", sort.getApiName());
        }
        return params;
    }

    public static DiscoverQuery getDiscoverQuery(Category category, int page) {
        if (category == null || category.categoryId == Category.ALL_CATEGORIES_ID) {
            return getDiscover(page);
        } else {
            return getDiscoverCategory(category.categoryId, page, SortTypes.MAGIC);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DiscoverQuery that = (DiscoverQuery) o;

        if (queryMap != null ? !queryMap.equals(that.queryMap) : that.queryMap != null) return false;
        return discoverType == that.discoverType;

    }

    @Override
    public int hashCode() {
        int result = queryMap != null ? queryMap.hashCode() : 0;
        result = 31 * result + discoverType.hashCode();
        return result;
    }
}

File: app/src/main/java/com/byoutline/kickmaterial/model/DiscoverQuery.kt
package com.byoutline.kickmaterial.model

import android.databinding.BaseObservable
import java.util.*

/**
 * @author Sebastian Kacprzak <sebastian.kacprzak at byoutline.com>
 */
class DiscoverQuery(val queryMap: Map<String, String>, val discoverType: DiscoverType) : BaseObservable() {

    val pageFromQuery: Int?
        get() {
            if (queryMap.containsKey("page")) {
                return Integer.valueOf(queryMap["page"])
            }
            return null
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as DiscoverQuery?

        if (queryMap != that!!.queryMap) return false
        return discoverType == that.discoverType

    }

    override fun hashCode(): Int {
        var result = queryMap.hashCode()
        result = 31 * result + discoverType.hashCode()
        return result
    }

    companion object {
        private const val PER_PAGE = 12

        private fun getDiscover(page: Int?): DiscoverQuery {
            val firstPage = page == null || page == 1
            val params = if (firstPage) emptyMap<String, String>() else getDiscoverCategoryMap(null, page, PER_PAGE, null)
            return DiscoverQuery(params, DiscoverType.DISCOVER)
        }

        private fun getDiscoverCategory(categoryId: Int, page: Int, sort: SortTypes): DiscoverQuery {
            val params = getDiscoverCategoryMap(categoryId, page, PER_PAGE, sort)
            return DiscoverQuery(params, DiscoverType.DISCOVER_CATEGORY)
        }

        fun getDiscoverSearch(searchTerm: String, categoryId: Int?, page: Int?, sort: SortTypes): DiscoverQuery {
            val params = getDiscoverCategoryMap(categoryId, page, PER_PAGE, sort)
            params.put("term", searchTerm)
            return DiscoverQuery(params, DiscoverType.SEARCH)
        }

        private fun getDiscoverCategoryMap(categoryId: Int?, page: Int?, perPage: Int?, sort: SortTypes?): MutableMap<String, String> {
            val params = HashMap<String, String>()
            if (categoryId != null) {
                params.put("category_id", Integer.toString(categoryId))
            }
            if (page != null) {
                params.put("page", Integer.toString(page))
            }
            if (perPage != null) {
                params.put("per_page", Integer.toString(perPage))
            }
            if (sort != null) {
                params.put("sort", sort.apiName)
            }
            return params
        }

        fun getDiscoverQuery(category: Category?, page: Int): DiscoverQuery {
            return if (category == null || category.categoryId == Category.ALL_CATEGORIES_ID) {
                getDiscover(page)
            } else {
                getDiscoverCategory(category.categoryId, page, SortTypes.MAGIC)
            }
        }
    }
}
